File: scripts/sync-bill-votes.ts
/**
 * Syncs bill vote summaries and member positions.
 * For each bill with recorded votes, fetches:
 *   - Senate votes: Senate.gov XML → parsed + lis_id resolved to bioguide_id
 *   - House votes:  Congress.gov House Roll Call API (119th Congress)
 *
 * Run manually: npx tsx scripts/sync-bill-votes.ts
 * Or triggered via: POST /api/cron/sync-bill-votes
 */

import { createServiceClient } from '@/lib/supabase/service'
import { parseSenateVoteXml } from './lib/parse-senate-vote-xml'
import { fetchHouseVote } from './lib/fetch-house-vote'
import { buildLisMap } from './lib/resolve-ids'

const CONGRESS_API_KEY = process.env.CONGRESS_API_KEY ?? ''
const CONGRESS_BASE = 'https://api.congress.gov/v3'
const DEFAULT_CONGRESS = 119
const BILLS_PER_PAGE = 20
// Look back this many days for bills with recent activity
const LOOKBACK_DAYS = 7

export interface SyncBillVotesResult {
  source: 'bill-votes'
  billsProcessed: number
  newVotes: number
  positionsStored: number
  senateLisResolutions: number
  senateResolutionFailures: number
  duration: string
}

interface RecordedVote {
  chamber: string
  congress: number
  date: string
  rollNumber: number
  sessionNumber: number
  url: string
}

interface BillRef {
  id: string
  billNumber: string  // e.g. "H.R. 1234"
  billTitle: string   // e.g. "The Example Act of 2025"
}

// Maps lowercase bill type codes to their standard abbreviated forms.
function formatBillNumber(type: string, number: string | number): string {
  const abbrevs: Record<string, string> = {
    hr:       'H.R.',
    s:        'S.',
    hjres:    'H.J.Res.',
    sjres:    'S.J.Res.',
    hres:     'H.Res.',
    sres:     'S.Res.',
    hconres:  'H.Con.Res.',
    sconres:  'S.Con.Res.',
  }
  const prefix = abbrevs[type.toLowerCase()] ?? type.toUpperCase()
  return `${prefix} ${number}`
}

async function congressFetch(path: string): Promise<any> {
  const sep = path.includes('?') ? '&' : '?'
  const res = await fetch(
    `${CONGRESS_BASE}${path}${sep}format=json&api_key=${CONGRESS_API_KEY}`
  )
  if (!res.ok) throw new Error(`Congress.gov ${path}: ${res.status}`)
  return res.json()
}

async function getBillsWithRecentActivity(congress: number, lookbackDays: number): Promise<BillRef[]> {
  const from = new Date(Date.now() - lookbackDays * 24 * 60 * 60 * 1000)
    .toISOString()
    .replace(/\.\d{3}Z$/, 'Z')

  const bills: BillRef[] = []
  let offset = 0

  while (true) {
    const data = await congressFetch(
      `/bill/${congress}?fromDateTime=${from}&limit=${BILLS_PER_PAGE}&offset=${offset}&sort=updateDate+desc`
    )
    const raw: any[] = data.bills ?? []
    for (const b of raw) {
      if (b.type && b.number) {
        bills.push({
          id:         `${congress}-${b.type.toLowerCase()}-${b.number}`,
          billNumber: formatBillNumber(b.type, b.number),
          billTitle:  b.title ?? '',
        })
      }
    }
    if (raw.length < BILLS_PER_PAGE) break
    offset += BILLS_PER_PAGE
    if (offset >= 200) break  // cap at 200 bills per sync run
  }

  return bills
}

async function getRecordedVotesForBill(billId: string): Promise<RecordedVote[]> {
  const [congress, type, number] = billId.split('-')
  const data = await congressFetch(
    `/bill/${congress}/${type}/${number}/actions?limit=50`
  )
  const actions: any[] = data.actions ?? []

  const votes: RecordedVote[] = []
  for (const action of actions) {
    for (const rv of action.recordedVotes ?? []) {
      if (rv.rollNumber && rv.url) {
        votes.push({
          chamber:       rv.chamber ?? '',
          congress:      rv.congress ?? parseInt(congress),
          date:          rv.date ?? action.actionDate ?? '',
          rollNumber:    rv.rollNumber,
          sessionNumber: rv.sessionNumber ?? 1,
          url:           rv.url,
        })
      }
    }
  }
  return votes
}

export async function syncBillVotes(
  congress: number = DEFAULT_CONGRESS,
  lookbackDays: number = LOOKBACK_DAYS
): Promise<SyncBillVotesResult> {
  if (!CONGRESS_API_KEY) throw new Error('CONGRESS_API_KEY not set')

  const start = Date.now()
  const supabase = createServiceClient()

  let billsProcessed = 0
  let newVotes = 0
  let positionsStored = 0
  let senateLisResolutions = 0
  let senateResolutionFailures = 0

  // Get bills with recent activity
  const billRefs = await getBillsWithRecentActivity(congress, lookbackDays)
  billsProcessed = billRefs.length

  for (const bill of billRefs) {
    const { id: billId, billNumber, billTitle } = bill

    let recordedVotes: RecordedVote[]
    try {
      recordedVotes = await getRecordedVotesForBill(billId)
    } catch {
      continue
    }

    for (const rv of recordedVotes) {
      const voteId = `${rv.congress}-${rv.chamber.toLowerCase()}-${rv.rollNumber}`

      // Skip if already synced; backfill title if it was never set
      const { data: existing } = await supabase
        .from('bill_vote_summaries')
        .select('id, title, question')
        .eq('id', voteId)
        .maybeSingle()
      if (existing) {
        if (!existing.title) {
          const voteTitle = billTitle
            ? `${billNumber}: ${billTitle} — ${existing.question}`
            : `${billNumber} — ${existing.question}`
          await supabase
            .from('bill_vote_summaries')
            .update({ title: voteTitle })
            .eq('id', voteId)
        }
        continue
      }

      try {
        if (rv.chamber.toLowerCase() === 'senate') {
          // ── Senate: fetch XML from senate.gov ─────────────────────────────
          const xmlRes = await fetch(rv.url)
          if (!xmlRes.ok) continue
          const xml = await xmlRes.text()
          const parsed = parseSenateVoteXml(xml)

          // Resolve lis_ids to bioguide_ids
          const lisIds = parsed.members.map(m => m.lis_member_id)
          const lisMap = await buildLisMap(supabase, lisIds)

          const voteTitle = billTitle
            ? `${billNumber}: ${billTitle} — ${parsed.question}`
            : `${billNumber} — ${parsed.question}`

          const { error: sumErr } = await supabase
            .from('bill_vote_summaries')
            .upsert({
              id:               voteId,
              bill_id:          billId,
              congress:         rv.congress,
              chamber:          'Senate',
              date:             rv.date,
              title:            voteTitle,
              question:         parsed.question,
              result:           parsed.result,
              yea_total:        parsed.yea_total,
              nay_total:        parsed.nay_total,
              present_total:    parsed.present_total,
              not_voting_total: parsed.not_voting_total,
              yea_democrat:     parsed.yea_democrat,
              nay_democrat:     parsed.nay_democrat,
              yea_republican:   parsed.yea_republican,
              nay_republican:   parsed.nay_republican,
              yea_independent:  parsed.yea_independent,
              nay_independent:  parsed.nay_independent,
              source_url:       rv.url,
              synced_at:        new Date().toISOString(),
            }, { onConflict: 'id' })
          if (sumErr) continue

          const positions: any[] = []
          for (const member of parsed.members) {
            const bioguide = lisMap.get(member.lis_member_id)
            if (bioguide) {
              positions.push({ vote_id: voteId, bioguide_id: bioguide, position: member.vote_cast })
              senateLisResolutions++
            } else {
              senateResolutionFailures++
            }
          }

          if (positions.length > 0) {
            const { error: posErr } = await supabase
              .from('bill_vote_positions')
              .upsert(positions, { onConflict: 'vote_id,bioguide_id' })
            if (!posErr) positionsStored += positions.length
          }

          newVotes++

        } else {
          // ── House: Congress.gov House Roll Call API ────────────────────────
          const houseData = await fetchHouseVote(rv.congress, rv.rollNumber, CONGRESS_API_KEY, rv.sessionNumber)
          if (!houseData) continue

          const voteTitle = billTitle
            ? `${billNumber}: ${billTitle} — ${houseData.question}`
            : `${billNumber} — ${houseData.question}`

          const { error: sumErr } = await supabase
            .from('bill_vote_summaries')
            .upsert({
              id:               voteId,
              bill_id:          billId,
              congress:         rv.congress,
              chamber:          'House',
              date:             rv.date,
              title:            voteTitle,
              question:         houseData.question,
              result:           houseData.result,
              yea_total:        houseData.yea_total,
              nay_total:        houseData.nay_total,
              present_total:    houseData.present_total,
              not_voting_total: houseData.not_voting_total,
              yea_democrat:     houseData.yea_democrat,
              nay_democrat:     houseData.nay_democrat,
              yea_republican:   houseData.yea_republican,
              nay_republican:   houseData.nay_republican,
              yea_independent:  houseData.yea_independent,
              nay_independent:  houseData.nay_independent,
              source_url:       rv.url,
              synced_at:        new Date().toISOString(),
            }, { onConflict: 'id' })
          if (sumErr) continue

          const positions = houseData.members.map(m => ({
            vote_id:     voteId,
            bioguide_id: m.bioguide_id,
            position:    m.position,
          }))

          if (positions.length > 0) {
            // Filter to only known legislators
            const bioguideIds = positions.map(p => p.bioguide_id)
            const { data: knownLeg } = await supabase
              .from('legislators')
              .select('bioguide_id')
              .in('bioguide_id', bioguideIds)
            const known = new Set((knownLeg ?? []).map((r: any) => r.bioguide_id))
            const validPositions = positions.filter(p => known.has(p.bioguide_id))

            if (validPositions.length > 0) {
              const { error: posErr } = await supabase
                .from('bill_vote_positions')
                .upsert(validPositions, { onConflict: 'vote_id,bioguide_id' })
              if (!posErr) positionsStored += validPositions.length
            }
          }

          newVotes++
        }
      } catch (err) {
        console.error(`Failed to sync vote ${voteId}:`, err)
        continue
      }
    }
  }

  const duration = `${((Date.now() - start) / 1000).toFixed(1)}s`
  return {
    source: 'bill-votes',
    billsProcessed,
    newVotes,
    positionsStored,
    senateLisResolutions,
    senateResolutionFailures,
    duration,
  }
}

if (require.main === module || process.argv[1]?.endsWith('sync-bill-votes.ts')) {
  syncBillVotes()
    .then(r => console.log(JSON.stringify(r, null, 2)))
    .catch(err => { console.error(err); process.exit(1) })
}

File: pipeline/scripts/sync_bill_votes.py
"""
Syncs bill vote summaries and member positions.

Run: python pipeline/scripts/sync_bill_votes.py

Options (env vars):
  CONGRESS_API_KEY  — required
  LOOKBACK_DAYS     — days of recent activity to look back (default: 7)
"""
import os
import sys
import time
import json
from datetime import datetime, timezone, timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent))

import httpx
from pipeline.lib.supabase_client import create_service_client
from pipeline.lib.parse_senate_vote_xml import parse_senate_vote_xml
from pipeline.lib.fetch_house_vote import fetch_house_vote
from pipeline.lib.resolve_ids import build_lis_map
from pipeline.lib.config import (
    CONGRESS_BASE, DEFAULT_CONGRESS, PAGE_SIZE_VOTES,
    LOOKBACK_DAYS_VOTES, TIMEOUT_DEFAULT, TIMEOUT_SHORT, BILLS_OFFSET_LIMIT,
    RATE_LIMIT_CONGRESS,
)

CONGRESS_API_KEY = os.environ.get("CONGRESS_API_KEY", "")
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", str(LOOKBACK_DAYS_VOTES)))

BILL_TYPE_ABBREVS: dict[str, str] = {
    "hr":      "H.R.",
    "s":       "S.",
    "hjres":   "H.J.Res.",
    "sjres":   "S.J.Res.",
    "hres":    "H.Res.",
    "sres":    "S.Res.",
    "hconres": "H.Con.Res.",
    "sconres": "S.Con.Res.",
}


def format_bill_number(bill_type: str, number: str | int) -> str:
    prefix = BILL_TYPE_ABBREVS.get(bill_type.lower(), bill_type.upper())
    return f"{prefix} {number}"


def congress_fetch(path: str) -> dict:
    sep = "&" if "?" in path else "?"
    resp = httpx.get(
        f"{CONGRESS_BASE}{path}{sep}format=json&api_key={CONGRESS_API_KEY}",
        timeout=TIMEOUT_DEFAULT,
    )
    resp.raise_for_status()
    return resp.json()


def get_bills_with_recent_activity(congress: int, lookback_days: int) -> list[dict]:
    from_dt = (datetime.now(timezone.utc) - timedelta(days=lookback_days)).strftime("%Y-%m-%dT%H:%M:%SZ")
    bills = []
    offset = 0

    while True:
        data = congress_fetch(
            f"/bill/{congress}?fromDateTime={from_dt}&limit={PAGE_SIZE_VOTES}&offset={offset}&sort=updateDate+desc"
        )
        raw: list[dict] = data.get("bills") or []
        for b in raw:
            if b.get("type") and b.get("number"):
                bills.append({
                    "id":         f"{congress}-{b['type'].lower()}-{b['number']}",
                    "bill_number": format_bill_number(b["type"], b["number"]),
                    "bill_title":  b.get("title", ""),
                })
        time.sleep(RATE_LIMIT_CONGRESS)
        if len(raw) < PAGE_SIZE_VOTES:
            break
        offset += PAGE_SIZE_VOTES
        if offset >= BILLS_OFFSET_LIMIT:
            break

    return bills


def get_recorded_votes_for_bill(bill_id: str) -> list[dict]:
    parts = bill_id.split("-")
    congress, bill_type, number = parts[0], parts[1], parts[2]
    data = congress_fetch(f"/bill/{congress}/{bill_type}/{number}/actions?limit=50")
    votes = []
    for action in data.get("actions") or []:
        for rv in action.get("recordedVotes") or []:
            if rv.get("rollNumber") and rv.get("url"):
                votes.append({
                    "chamber":        rv.get("chamber", ""),
                    "congress":       rv.get("congress") or int(congress),
                    "date":           rv.get("date") or action.get("actionDate", ""),
                    "roll_number":    rv["rollNumber"],
                    "session_number": rv.get("sessionNumber", 1),
                    "url":            rv["url"],
                })
    return votes


def sync_bill_votes(congress: int = DEFAULT_CONGRESS, lookback_days: int = LOOKBACK_DAYS) -> dict:
    if not CONGRESS_API_KEY:
        raise RuntimeError("CONGRESS_API_KEY not set")

    start = time.time()
    supabase = create_service_client()
    now = datetime.now(timezone.utc).isoformat()

    bills_processed = new_votes = positions_stored = 0
    senate_lis_resolutions = senate_resolution_failures = 0

    bill_refs = get_bills_with_recent_activity(congress, lookback_days)
    bills_processed = len(bill_refs)

    for bill in bill_refs:
        bill_id = bill["id"]
        bill_number = bill["bill_number"]
        bill_title = bill["bill_title"]

        try:
            recorded_votes = get_recorded_votes_for_bill(bill_id)
        except Exception:
            continue
        time.sleep(RATE_LIMIT_CONGRESS)

        for rv in recorded_votes:
            vote_id = f"{rv['congress']}-{rv['chamber'].lower()}-{rv['roll_number']}"

            # Skip if already synced; backfill title if missing
            existing = (
                supabase.table("bill_vote_summaries")
                .select("id, title, question")
                .eq("id", vote_id)
                .maybe_single()
                .execute()
            )
            if existing.data:
                if not existing.data.get("title"):
                    q = existing.data.get("question", "")
                    vote_title = f"{bill_number}: {bill_title} — {q}" if bill_title else f"{bill_number} — {q}"
                    supabase.table("bill_vote_summaries").update({"title": vote_title}).eq("id", vote_id).execute()
                continue

            try:
                if rv["chamber"].lower() == "senate":
                    # Senate: fetch XML from senate.gov
                    xml_resp = httpx.get(rv["url"], timeout=TIMEOUT_SHORT)
                    if not xml_resp.is_success:
                        continue
                    parsed = parse_senate_vote_xml(xml_resp.text)

                    # Resolve lis_ids to bioguide_ids
                    lis_ids = [m.lis_member_id for m in parsed.members]
                    lis_map = build_lis_map(supabase, lis_ids)

                    vote_title = (
                        f"{bill_number}: {bill_title} — {parsed.question}"
                        if bill_title else
                        f"{bill_number} — {parsed.question}"
                    )

                    sum_result = supabase.table("bill_vote_summaries").upsert({
                        "id":               vote_id,
                        "bill_id":          bill_id,
                        "congress":         rv["congress"],
                        "chamber":          "Senate",
                        "date":             rv["date"],
                        "title":            vote_title,
                        "question":         parsed.question,
                        "result":           parsed.result,
                        "yea_total":        parsed.yea_total,
                        "nay_total":        parsed.nay_total,
                        "present_total":    parsed.present_total,
                        "not_voting_total": parsed.not_voting_total,
                        "yea_democrat":     parsed.yea_democrat,
                        "nay_democrat":     parsed.nay_democrat,
                        "yea_republican":   parsed.yea_republican,
                        "nay_republican":   parsed.nay_republican,
                        "yea_independent":  parsed.yea_independent,
                        "nay_independent":  parsed.nay_independent,
                        "source_url":       rv["url"],
                        "synced_at":        now,
                    }, on_conflict="id").execute()

                    positions = []
                    for member in parsed.members:
                        bioguide = lis_map.get(member.lis_member_id)
                        if bioguide:
                            positions.append({
                                "vote_id":     vote_id,
                                "bioguide_id": bioguide,
                                "position":    member.vote_cast,
                            })
                            senate_lis_resolutions += 1
                        else:
                            senate_resolution_failures += 1

                    if positions:
                        supabase.table("bill_vote_positions").upsert(
                            positions, on_conflict="vote_id,bioguide_id"
                        ).execute()
                        positions_stored += len(positions)

                    new_votes += 1

                else:
                    # House: Congress.gov House Roll Call API
                    house_data = fetch_house_vote(
                        rv["congress"], rv["roll_number"], CONGRESS_API_KEY, rv["session_number"]
                    )
                    if not house_data:
                        continue

                    vote_title = (
                        f"{bill_number}: {bill_title} — {house_data.question}"
                        if bill_title else
                        f"{bill_number} — {house_data.question}"
                    )

                    supabase.table("bill_vote_summaries").upsert({
                        "id":               vote_id,
                        "bill_id":          bill_id,
                        "congress":         rv["congress"],
                        "chamber":          "House",
                        "date":             rv["date"],
                        "title":            vote_title,
                        "question":         house_data.question,
                        "result":           house_data.result,
                        "yea_total":        house_data.yea_total,
                        "nay_total":        house_data.nay_total,
                        "present_total":    house_data.present_total,
                        "not_voting_total": house_data.not_voting_total,
                        "yea_democrat":     house_data.yea_democrat,
                        "nay_democrat":     house_data.nay_democrat,
                        "yea_republican":   house_data.yea_republican,
                        "nay_republican":   house_data.nay_republican,
                        "yea_independent":  house_data.yea_independent,
                        "nay_independent":  house_data.nay_independent,
                        "source_url":       rv["url"],
                        "synced_at":        now,
                    }, on_conflict="id").execute()

                    # Filter to known legislators
                    all_positions = [
                        {"vote_id": vote_id, "bioguide_id": m.bioguide_id, "position": m.position}
                        for m in house_data.members
                    ]
                    if all_positions:
                        bioguide_ids = [p["bioguide_id"] for p in all_positions]
                        known_resp = (
                            supabase.table("legislators")
                            .select("bioguide_id")
                            .in_("bioguide_id", bioguide_ids)
                            .execute()
                        )
                        known = {r["bioguide_id"] for r in (known_resp.data or [])}
                        valid_positions = [p for p in all_positions if p["bioguide_id"] in known]
                        if valid_positions:
                            supabase.table("bill_vote_positions").upsert(
                                valid_positions, on_conflict="vote_id,bioguide_id"
                            ).execute()
                            positions_stored += len(valid_positions)

                    new_votes += 1

            except Exception as err:
                print(f"  Failed to sync vote {vote_id}: {err}")
                continue
            finally:
                time.sleep(RATE_LIMIT_CONGRESS)

    duration = f"{time.time() - start:.1f}s"
    return {
        "source": "bill-votes",
        "bills_processed": bills_processed,
        "new_votes": new_votes,
        "positions_stored": positions_stored,
        "senate_lis_resolutions": senate_lis_resolutions,
        "senate_resolution_failures": senate_resolution_failures,
        "duration": duration,
    }


if __name__ == "__main__":
    result = sync_bill_votes()
    print(json.dumps(result, indent=2))
